File: remove-invalid-parenthesis.py
#Time complexity : O(N.2^N) (For isValid and 2^N possible strings)
#Space complexity: O(N.2^N) 


class Solution:
    def removeInvalidParentheses(self, s: str) -> List[str]:
        def isValid(s):
            count = 0
            for char in s:
                if char == '(':
                    count += 1
                elif char == ')':
                    count -= 1
                if count < 0:
                    return False
            return count == 0

        result = []
        seen = set([s])
        queue = deque([s])
        found = False

        while queue:
            for _ in range(len(queue)):
                curr = queue.popleft()
                if isValid(curr):
                    result.append(curr)
                    found = True
                if found:
                    continue # Stop if valid string found at current level

                # Generate all possible states
                for i in range(len(curr)):
                    if curr[i] in "()": # Only consider parentheses
                        next_str = curr[:i] + curr[i+1:]
                        if next_str not in seen:
                            seen.add(next_str)
                            queue.append(next_str)

            if found:
                break # Stop if valid strings have been found at this level

        return result

File: removeInvalidParenthesis.cpp
/*Intuition: Try to remove the bracket at each an every index. Be exhaustive. Use BFS so that we can find the answer at the topmost level.
Time: O(N^N)
Space: O(N^N)
*/

class Solution {
public:
    vector<string> removeInvalidParentheses(string s) {
        vector<string> result;
        queue<string>queue;
        queue.push(s);
        set<string>visited;
        bool flag = false;
        visited.insert(s);
        while ( queue.size() != 0){
            
            int size = queue.size();
            for ( int i = 0; i < size; i++){
            
                
                auto currString = queue.front();
                queue.pop();
                
                if ( isValid(currString)){
                    flag = true;
                    result.push_back(currString);
                
                }
                if ( flag == false){
                
                
                    for ( int k = 0; k < currString.size(); k++){
                        if ( isalpha(currString[k])) continue;
                        string newString = currString.substr(0,k) + currString.substr(k+1);
                        if ( visited.find(newString) == visited.end()){
                            queue.push(newString);
                            visited.insert(newString);
                        }

                    }
                }
                
            
            
            }
        
        }
        return result;
        
        
    }
    bool isValid(string s){
        int count = 0;
        for ( int i = 0; i < s.size(); i++){
            if ( s[i] == ')'){
                if ( count == 0) return false;
                count--;
            
            }
            else if(s[i] == '('){
                count++;
            }
        }
        return count == 0;
    
    }
};